api/circulation.py

```diff
@@ -459,6 +459,65 @@ def can_revoke_hold(self, licensepool, hold):
             return True
         return False
 
+    def _collect_event(self, patron, licensepool, name,
```

Comment:
There are a lot of rules in `fulfill` about where to get `patron` and `library` under different scenarios. That's because you can fulfill a book without authenticating (this happens in The SimplyE Collection(tm) for example). I centralized rules for analytics collection in this method, so that I could centralize the rules about pulling the 'neighborhood' out of the request.

However, the rules for finding the 'neighborhood' depend on having `patron` first. It was easiest to _also_ move in the complicated rules for finding `patron` and `library` in various places, and run that logic all the time, even though `fulfill` is the only time we currently need that functionality.